### DISCUSSION TITLE
"no index.html found"

### DISCUSSION THREAD

**Discussion by DonFuego** [wailsapp/wails#2096] [General]
Hi, we are testing an application build on both mac and linux.  After we run `wails build` and attempt to launch the program from the `build/bin` directory we get the "no index.html found" error on both operating systems.  I'm not sure if it's an application bug or not.  It's likely that it is an application bug...can anyone point me in the direction to attempt to solve?  What settings or build steps should I be looking at?  

Any help appreciated.

**Comment by leaanthony** [comment]
Did you set the assets field? It should warn you if you didn't so that might be on us.

**Reply by stffabi** [reply]
Please also make sure the embedded directory contains your frontend assets, see also [here](https://wails.io/docs/guides/troubleshooting#my-application-is-displaying-a-whiteblank-screen).

There are use cases when a developer wants not to set the assets, for example if they want to use only the assetshandler. So we could probably warn if none of both has been set.

What we currently do is warn in dev mode if someone hasn't set the assets but is using a frontenddevserver. In such case we also disable the frontenddevserver, so at least then also dev mode doesn't display content.

**Reply by stffabi** [reply]
If you see `2022/11/16 07:39:55 no index.html found` in the console and the app stops => The assets have been set, but the fs.FS doesn't contain any index.html, which would indicate the wrong directory has been embedded. I'll provide a PR to improve the error message in this case to: `no 'index.html' could be found in your Assets fs.FS, please make sure the embedded directory 'frontend/src' is correct and contains your assets`

If the app starts and you see a website with the wails logo indicating the index.html could not be found: This means there were not assets set and also no assetshandler.

**Reply by leaanthony** [reply]
I think the error page could give all the information including a link to the troubleshooting guide

**Reply by stffabi** [reply]
> I think the error page could give all the information including a link to the troubleshooting guide

Yeah that's also possible, would be a little bit more of work. Maybe we should create a issue for that. Improving the current error messages was quite straight forward and is done. @leaanthony do you want to already merge that for the time being? That would at least improve it until we have a customised error page with different error messages.

**Reply by leaanthony** [reply]
Absolutely! Thanks so much for doing this 🙏

**Reply by DonFuego** [reply]
Hi guys, thank you so much for jumping in here with some help.  A few questions: 

1. Where do I set the assets field?  
2. I'm using svelte-kit instead of standard svelte as I wanted to take advantage of routing/templating/etc. that comes out of the box.  I'm sure that setup is causing some issues.
3. Due to svelte-kit, my root html is actually called app.html and not index.html.  However, I updated the svelte config to look for that as the app template and I can still run `wails dev` just fine.  But thought I'd point this part out. 
4. I did look at the `frontend/dist` directory and it appears to be empty with just a `gitkeep` file in there.  Is this not where the assets _should_ be served from?  Should I change the embedded directive to `frontend/src`?

I think we're close - thanks for your patience.  Still understanding how all this works.

**Reply by DonFuego** [reply]
So I just tried changing the embed directive to `frontend/src` and `wails dev` appears to be serving up the svelte application just fine.  However, after running `wails build` then launching the application I'm getting a blank application with just my svelte root index page simply showing `%sveltekit.head% %sveltekit.body%` and not rendering.

In the terminal, I just see:
```
Overriding existing handler for signal 10. Set JSC_SIGNAL_FOR_GC if you want WebKit to use a different signal

(WebKitWebProcess:857791): GLib-CRITICAL **: 08:44:51.978: Failed to set scheduler settings: Operation not permitted
```

**Reply by DonFuego** [reply]
What else I find odd, is it appears that no matter what I set the embed directive too: `frontend/dist` or `frontend/src` the application seems to run just fine in `wails dev`

**Reply by stffabi** [reply]
There are quite some differences between `wails dev` and a production build. Wails dev takes benefit of a frontend dev server like Vite, to support fast reloading and improve the developer experience. As for a production build that server can't be used and the frontend needs to packaged and embedded into the binary.

I suspect there's some missing configuration to package your frontend. Your `go:embed` directive must point to the folder where your frontend build process puts the finalised assets. During a `wails build` the `"frontend:install"` and `"frontend:build"` (those can be found in your wails.json) commands are used to package the frontend. 
Try to run those commands in your `"frontend:dir"` directory, they should build e.g. a `dist` folder with all generated output. That's the folder that you need to embed in `main.go`. You can also try to open that folder in a normal browser and navigate to the `index.html`. If it shows your desired frontend, it should also work in Wails.

**Reply by stffabi** [reply]
Hopefully this will help you to narrow down the problem.

**Reply by DonFuego** [reply]
That definitely helps - sounds like I need to do some digging on how this is getting packaged up on the front-end for a build.  Thank you for the prompt response!

**Reply by DonFuego** [reply]
Another question - if there are two different folders where the assets live: 1) during development in `frontend/src` and 2) during build time in a separate folder somewhere else say `.svelte-kid/build`.  How would you recommend switching between those during development short of just manually updating the go embed directive prior to doing dev vs. build?

**Reply by DonFuego** [reply]
or am I misunderstanding how this is put together - is the `go:embed` _ONLY_ used during the build/bundling phase?

**Reply by DonFuego** [reply]
@leaanthony @stffabi I Just wanted to report that I was able to successfully get it to build and run on mac and linux.  I had to tweak the svelte config and update the go embed to point to `frontend/build`.  Thank you guys for your help!
